### ISSUE TITLE
Rule idea: at-(extend/include)-empty-line-(before/after)

### ISSUE THREAD

**Issue by jrencz** [kristerkari/stylelint-scss#67]
One may want to visually separate includes, properties (or nestings) and extends (in whatever order they come in, relates to #43)

I suggest a rule which will allow enforcing a single new line before extend, let's call it `at-extend-empty-line-before`:

```scss
// bad
a {
  width: 10px;
  @extend %some-placeholder;
}

// good
a {
  width: 10px;
  
  @extend %some-placeholder;
}
```

There might be an additional option to disallow new line inside group of `@extend`s:
```scss
// with described optional option, bad
a {
  width: 10px;
  
  @extend %some-placeholder;
  
  @extend %some-other-placeholder;
}

// without described optional option: good
a {
  width: 10px;
  
  @extend %some-placeholder;
  @extend %some-other-placeholder;
}
```

I just don't have an idea how to call it. I base the idea on [`declaration-empty-line-before` except: "after-declaration"](http://stylelint.io/user-guide/rules/declaration-empty-line-before/#after-declaration), [`declaration-empty-line-before` except: ["first-nested"]](http://stylelint.io/user-guide/rules/declaration-empty-line-before/#first-nested) and [`rule-nested-empty-line-before` except: ["first-nested"]](http://stylelint.io/user-guide/rules/rule-nested-empty-line-before/#except-first-nested). I think it might be default.

The rule may come in pair with `at-extend-empty-line-after` for those who prefer extends at the top of the block (which is somewhat related to #43) yet they would like to separate them.

The rule may also come in pair with `at-include-empty-line-before`/`at-include-empty-line-after` for `@include`:

```scss
// bad
a {
  @include some-mixin;
  width: 10px;
}

// good
a {
  @include some-mixin;
  
  width: 10px;
}
```

```scss
// with described optional option, bad
a {
  @include some-mixin;

  @include some-other-mixin;
  
  width: 10px;
}

// without described optional  option: good
a {
  @include some-mixin;
  @include some-other-mixin;
  
  width: 10px;
}
```

**Comment by jeddy3** [comment]
@jrencz Thanks for the detailed write-up. I'm not don't contribute to `postcss-scss`, but I do contribute to `stylelint` itself. Perhaps I can help here.

I think what you're after can already be achieved using stylelint's core [`at-rule-empty-line-before`](http://stylelint.io/user-guide/rules/at-rule-empty-line-before/) rule.

With the following config:

```json
{
  "rules": {
    "at-rule-empty-line-before": [ "always", {
      "except": [
        "blockless-after-same-name-blockless",
        "first-nested"
      ]
    } ],
    "declaration-empty-line-before": [ "always", {
      "except": [
        "after-declaration",
        "first-nested"
      ]
    } ],
  }
}
```

The following is good:

```scss
a {
  @include some-mixin;
  @include some-mixin;
  @extend %some-placeholder;

  width: 10px;
}
```

But the following will warn:

```scss
a {

  @include some-mixin;

  @include some-mixin;
  @extend %some-placeholder;

  width: 10px;
}
```

If you always want an empty line between your `@include`s and `@extend`s, change the `"blockless-after-same-name-blockless"` option to `"blockless-group"`, e.g. this would be then be valid:

```scss
a {
  @include some-mixin;
  @include some-mixin;

  @extend %some-placeholder;

  width: 10px;
}
```

**Comment by dryoma** [comment]
@jrencz does the configs from @jeddy3's post suit your need (i.e. can we close this)?

**Comment by jrencz** [comment]
Ah, I forgot to check then and right now I'm not able to check quickly. But conceptually it should work. If I check and I won't be able to achieve what I want I'll reopen this one or file a new one
